kzg_prover/src/circuits/utils.rs

```diff
@@ -192,82 +227,104 @@ pub fn verify_grand_sum_openings<const N_CURRENCIES: usize>(
         }
     }
 
-    let mut verification_results = Vec::<bool>::new();
-    let mut constant_terms = Vec::<BigUint>::new();
-
-    for (i, advice_commitment) in advice_commitments.iter().enumerate() {
-        let (verified, constant_term) = verify_opening::<
-            KZGCommitmentScheme<Bn256>,
-            VerifierSHPLONK<'_, Bn256>,
-            Challenge255<G1Affine>,
-            Blake2bRead<_, _, Challenge255<_>>,
-            AccumulatorStrategy<_>,
-        >(
-            params,
-            &challenge_opening_proofs[i],
-            Fp::zero(),
-            *advice_commitment,
-        );
-        verification_results.push(verified);
-
-        if verified {
-            constant_terms.push(fp_to_big_uint(constant_term * Fp::from(polynomial_degree)));
-        } else {
-            constant_terms.push(BigUint::from(0u8));
-        }
-    }
-    (verification_results, constant_terms)
+    let (verified, constant_terms) = verify_opening::<
+        KZGCommitmentScheme<Bn256>,
+        VerifierSHPLONK<'_, Bn256>,
+        Challenge255<G1Affine>,
+        Blake2bRead<_, _, Challenge255<_>>,
+        AccumulatorStrategy<_>,
+        N_CURRENCIES,
+    >(
+        params,
+        &grand_sum_opening_batch_proof,
+        Fp::zero(),
+        &advice_commitments,
+    );
+
+    (
+        verified,
+        constant_terms
+            .iter()
+            .map(|eval| fp_to_big_uint(eval * Fp::from(polynomial_degree)))
+            .collect(),
+    )
 }
 
-pub fn verify_user_inclusion<const N_CURRENCIES: usize>(
+/// Verifies the KZG batch proof of the polynomial openings being the evaluations
+/// of the advice polynomials at the point corresponding to the user index
+///
+/// # Arguments
+/// * `params` - the KZG parameters
+/// * `zk_snark_proof` - the ZK-SNARK proof of the circuit whose advice columns contain the user entry polynomials
+/// * `balance_opening_batch_proof` - the KZG batch proof of the user entry polynomials
+/// * `column_range` - the range of the advice columns that represent user entry
+/// * `omega` - $\omega$, the generator of the $2^k$ order multiplicative subgroup used to interpolate the polynomials.
+/// * `user_index` - the index of the user whose entry is being proven
+/// * `N_POINTS` - the size of the user entry being verified (e.g., 1 ID value + 4 balance values = 5)
+///
+/// # Returns
+/// * `bool` - whether the user entry openings are verified correctly
+/// * `Vec<BigUint>` - the evaluations of the advice polynomials at the point corresponding to the user index
+pub fn verify_user_inclusion<const N_POINTS: usize>(
     params: &ParamsKZG<Bn256>,
     zk_snark_proof: &[u8],
-    balance_opening_proofs: Vec<Vec<u8>>,
-    balance_column_range: Range<usize>,
+    balance_opening_batch_proof: &[u8],
+    column_range: Range<usize>,
```

Comment:
Could we use the range derived from `user_index` and `N_POINTS` instead of using `column_range`?

Reply:
No. User index is an index "along" the polynomial, which means a row index of advice columns. N_POINTS are the number of KZG commitments we're dealing with, but we cannot be sure where it starts. For example, if later we allocate the initial advice columns for something other that user ID, that would move the advice columns of interest to the right. I tried to make the function API general enough to be independent of our circuit implementation (because it also lives in a different module and is circuit-agnostic).

Reply:
I understood, thanks for explanation.